src/signal.rs

```diff
@@ -1,18 +1,132 @@
-use futures::Stream;
+use tokio::sync::{broadcast, mpsc};
+use tokio_stream::{Stream, StreamExt};
 
-#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub type ShutdownTx = broadcast::Sender<()>;
+pub type SignalTx = mpsc::Sender<SignalTo>;
+pub type SignalRx = mpsc::Receiver<SignalTo>;
+
+#[derive(Debug)]
+/// Control messages used by Vector to drive topology and shutdown events.
 pub enum SignalTo {
-    /// Signal to reload config.
-    Reload,
+    /// Signal to reload config from a string.
+    ReloadFromString(String),
+    /// Signal to reload config from the filesystem.
+    ReloadFromDisk,
     /// Signal to shutdown process.
     Shutdown,
     /// Shutdown process immediately.
     Quit,
 }
 
+/// SignalHandler is a general `ControlTo` message receiver and transmitter. It's used by
+/// OS signals and providers to surface control events to the root of the application.
+pub struct SignalHandler {
+    tx: SignalTx,
+    rx: Option<SignalRx>,
+    shutdown_txs: Vec<ShutdownTx>,
+}
+
+impl SignalHandler {
+    /// Create a new signal handler. We'll have space for 2 control messages at a time, to
+    /// ensure the channel isn't blocking.
+    pub fn new() -> Self {
+        let (tx, rx) = mpsc::channel(2);
+
+        Self {
+            tx,
+            rx: Some(rx),
+            shutdown_txs: vec![],
+        }
+    }
+
+    /// Clones the transmitter.
+    pub fn clone_tx(&self) -> SignalTx {
+        self.tx.clone()
+    }
+
+    /// Takes a stream who's elements are convertible to `SignalTo`, and spawns a permanent
+    /// task for transmitting to the receiver.
+    pub fn add<T, S>(&mut self, stream: S)
+    where
+        T: Into<SignalTo> + Send + Sync,
+        S: Stream<Item = T> + 'static + Send + Sync,
+    {
+        let tx = self.tx.clone();
+
+        tokio::spawn(async move {
```

Comment:
Rather than this implementation that spawns a whole new task, could we just have `SignalTx` be an actual wrapper struct with that does the conversion inline with sending?

Reply:
I've slightly tweaked this, although it still takes a stream and spawns a task to do the actual transmission. I think this is okay but please let me know if you still feel changes are needed here.